### ISSUE TITLE
GeoUtils or Rasterio doesn't close files properly.

### ISSUE THREAD

**Issue by erikmannerfelt** [GlacioHack/GeoUtils#153]
**Describe the bug**
`Raster`'s don't get unloaded when they go out of scope. I don't know if this is a Rasterio bug or GeoUtils, but it should definitely be solved within GeoUtils somehow.

**To Reproduce**

```python
import resource

import matplotlib.pyplot as plt
import geoutils as gu


def read_data():
    """Read a value from the data and return it."""
    r = gu.Raster(gu.datasets.get_path("landsat_B4"))
    return r.data[0, 0, 0]


ram = []
for _ in range(10000):
    read_data()
    ram.append(resource.getrusage(resource.RUSAGE_SELF).ru_maxrss)


plt.plot(ram)
plt.ylabel("RAM usage (KiB)")
plt.xlabel("Iterations")
plt.show()
```
![Screenshot_20210412_124248](https://user-images.githubusercontent.com/33550973/114382390-999f1b00-9b8c-11eb-9c19-e1f1061bc102.png)

EDIT: The memory magnitude is off, probably by `4096` since I didn't consider the page size. The `resource` calls are honestly over my head!

**Expected behavior**
RAM usage should not increase linearly by that amount (only by the amount equivalent to the RAM usage statistics). The dataset should be close once it goes out of scope in `read_data()`.

**System (please complete the following information):**
 - OS: Linux
 - Environment: Latest source

**Comment by erikmannerfelt** [comment]
This could perhaps be implemented with the `__del__(self)` method? Calling `.close()` or whatever it is called in the associated `DatasetReader` object maybe.

**Comment by atedstone** [comment]
I've had some success at replicating this behaviour locally, but I'm not running into real memory/performance issues.

System:
* OS: OSX (High Sierra, MBP, early 2015)
* geoutils at commit be186d8c1 (30 June 2021). Python 3.9.1, IPython 7.24.1, rasterio 1.2.0.

I re-ran the code snippet provided by @erikmannerfelt . It yields this graph:

<img width="591" alt="image" src="https://user-images.githubusercontent.com/2922144/124285237-08699000-db4e-11eb-8fd2-140f0b37f3f5.png">

Next I tried working with a collection of 18 GeoTIFFs, each 20-30 Mb in size. I ran the loop 300 times, picking an image to load at random from the collection each time. I haven't provided the images here, but here is the code:

```python 
import resource

import matplotlib.pyplot as plt
import geoutils as gu
import os
import psutil

process = psutil.Process(os.getpid())


def read_data(f):
    """Read a value from the data and return it."""
    #print('\tMax  Memory usage: %s (KB)' % resource.getrusage(resource.RUSAGE_SELF).ru_maxrss)
    #print('\tCurr Memory usage: %s (KB)' % (process.memory_info().rss / 1024))
    #r = gu.Raster(gu.datasets.get_path("landsat_B4"))
    r = gu.Raster(f)
    r.load()
    return r.data[0, 500:550, 500:550]

from glob import glob
files = glob('*.tif')
import random

ram = []
ram2 = []
for n in range(300):
    print(n)
    read_data(files[random.randrange(0,len(files))])
    #print('Max  Memory usage: %s (KB)' % resource.getrusage(resource.RUSAGE_SELF).ru_maxrss)
    #print('Curr Memory usage: %s (KB)' % (process.memory_info().rss / 1024))
    ram.append(resource.getrusage(resource.RUSAGE_SELF).ru_maxrss)
    ram2.append((process.memory_info().rss / 1024))

plt.figure()
ax = plt.subplot(111)
plt.plot(ram)
ax.ticklabel_format(useOffset=False, style='plain')
plt.ylabel('Max memory usage')

plt.figure()
ax = plt.subplot(111)
plt.plot(ram2)
plt.ylabel("Current RAM usage (KiB)")
plt.xlabel("Iterations")
ax.ticklabel_format(useOffset=False, style='plain')
plt.show()
```

I didn't see any real evidence of resource gobbling:

<img width="590" alt="image" src="https://user-images.githubusercontent.com/2922144/124286092-e15f8e00-db4e-11eb-875b-243fc53f56da.png">

I also watched the Python process in OSX Activity Monitor. Memory usage went up and down as the loop progressed. At the end of the loop the process was using about 200 Mb RAM, at most 50 Mb than it started with.

**Comment by atedstone** [comment]
Perhaps others, e.g. @adehecq , could also do some profiling their end, or at least keep a note of the steps and environments which lead to memory problems, then report back here?

**Comment by adehecq** [comment]
Thanks for looking into this! I will definitely have a look whenever I find time (is 2022 too late?? ;-))

**Comment by atedstone** [comment]
This will hopefully be solved by progress on #246 ...

**Comment by adehecq** [comment]
I tried running the same code snippet since the memory file removal in PR #265. I can still reproduce @erikmannerfelt's error... So it's difficult to assess whether or not we improved memory usage.

However, I wonder if this specific example is really an error, or a problem of how memory is handled in Python. I tried increasing the number of iterations or number of files read. At some point, the memory is capped if estimated with `resources`, or even goes down and up again if estimated with `psutils`. See figure below.
This was generated with the code below (note that I was also saving reading time, to check another issue I've encountered).

```
import resource
from time import time

import matplotlib.pyplot as plt
import geoutils as gu
from tqdm import tqdm

import psutil
import os
process = psutil.Process(os.getpid())

def read_data():
    """Read a value from the data and return it."""
    r = gu.Raster(gu.examples.get_path("everest_landsat_b4"))
    r2 = gu.Raster(gu.examples.get_path("everest_landsat_rgb"))
    r3 = gu.Raster(gu.examples.get_path("everest_landsat_b4_cropped"))
    return r.data[0, 0, 0]


ram = []
ram2 = []
proc_time = []
for _ in tqdm(range(50000)):
    t1 = time()
    res = read_data()
    proc_time.append(time() - t1)
    ram.append(resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / 1024)
    ram2.append((process.memory_info().rss / 1024))

plt.plot(ram, label="resource")
plt.plot(ram2, label="psutil")
plt.ylabel("RAM usage (KiB)")
plt.xlabel("Iterations")
plt.legend()

plt.figure()
plt.plot(proc_time)
plt.ylabel("Time (s)")
plt.xlabel("Iterations")
plt.legend()
plt.show()
```

![Figure_1](https://user-images.githubusercontent.com/3285905/198612963-8ed0b832-41bf-44d4-a5a6-26dfdb3eb8d0.png)

N.B: With some relatively large files, I noticed that loading time was 3s, and then after some time, loading exactly the same file took 12s. I could not find a MWE, because it seems to just happen at some point without any reason. Anyway, I am wondering if this can be related to the memory issue. I could not reproduce the issue with this example, as the processing time is rather constant over time (with some slight fluctuations, but not systematic).
